Lr scheduler

Jiaming-Liu opened the pull request:
Providing a unified LR scheduler.

Currently supports:
 - ReduceLROnPlateau (ported from [Keras](https://keras.io/))
 - LambdaLR
 - StepLR
 - MultiStepLR
 - ExponentialLR
 - GroupLambdaLR (Need testing)

Jiaming-Liu committed:
Update README.md

Jiaming-Liu committed:
Create lr_scheduler.py

Jiaming-Liu committed:
Update README.md

Jiaming-Liu committed:
Update README.md

Jiaming-Liu committed:
Update README.md

Jiaming-Liu committed:
Update README.md

Jiaming-Liu committed:
Update README.md

Jiaming-Liu committed:
Update README.md

Jiaming-Liu committed:
Update README.md

Jiaming-Liu committed:
Update lr_scheduler.py

Add 4 classes. Not tested.

Jiaming-Liu committed:
Update README.md

Jiaming-Liu committed:
Update lr_scheduler.py

Fix description.

Jiaming-Liu committed:
Update README.md

Jiaming-Liu committed:
Update lr_scheduler.py

Add MultiStepLR

Jiaming-Liu committed:
Update lr_scheduler.py

Jiaming-Liu committed:
Update .gitignore

Jiaming-Liu committed:
Update README.md

Jiaming-Liu committed:
Update .gitignore

Jiaming-Liu committed:
Revert README.md for merging

Jiaming-Liu committed:
Update lr_scheduler.py

Solving:
 - W293 blank line contains whitespace
 - W291 trailing whitespace

Jiaming-Liu committed:
Revert "Revert README.md for merging"

This reverts commit a74ec0d5aa8a3b5c4b5e78a411b67c9706befdbc.

Jiaming-Liu committed:
Revert "Revert "Revert README.md for merging""

This reverts commit 6df0a3435360c85e45c056c616b258780eef6d91.

colesbury reviewed (commented):
This looks generally good. I've added some inline comments.

This needs some unit tests. Basically:

- create an optimizer and LR scheduler
- for a few different values of 'epoch', call step on the scheduler and check that the LR of the optimizer is correct

Jiaming-Liu commented on torch/optim/lr_scheduler.py:
```diff
@@ -0,0 +1,181 @@
+        if self.mode not in ['min', 'max']:
+            raise RuntimeError('Learning Rate Plateau Reducing mode %s is unknown!')
+        if self.mode == 'min':
+            self.monitor_op = lambda a, b: np.less(a, b - self.epsilon)
```
ReduceLROnPlateau is ported from Keras. But yes, this sounds reasonable and I will make an option for it.

Jiaming-Liu replied:
@colesbury I am not sure about this one. Epsilon should be used for handling numerical error. If we really need a threshold, we might use something else.

Jiaming-Liu commented on torch/optim/lr_scheduler.py:
```diff
@@ -0,0 +1,181 @@
+        self.wait = 0
+        self.lr_epsilon = self.min_lr * 1e-4
+
+    def reset(self):
```
This part is added by mistake. Reset should not be called externally.

Jiaming-Liu commented on torch/optim/lr_scheduler.py:
```diff
@@ -0,0 +1,181 @@
+        self.zip = zip(optimizer.param_groups, base_lrs, lr_lambdas)
+
+    def step(self, epoch):
+        for param_group, base_lr, lr_lambda in self.zip:
```
This actually works fine in my testing cases. zip() produces a list.

colesbury replied:
Not in Python 3

Jiaming-Liu commented:
Not sure where to put the unit tester. I have put it [here](https://github.com/Jiaming-Liu/pytorch/blob/test-scheduler/test_lr_scheduler.py).

Jiaming-Liu committed:
Fix GroupLambdaLR

Jiaming-Liu committed:
Minor changes

Jiaming-Liu committed:
create _MonitorOp

To pack some details of ReduceLROnPlateau

Jiaming-Liu committed:
Minor changes

Rename _MonitorOp.best to _MonitorOp.worse

Jiaming-Liu committed:
Relative Threshold

Jiaming-Liu committed:
minor changes

Jiaming-Liu committed:
Fix E231

apaszke commented on torch/optim/lr_scheduler.py:
```diff
@@ -0,0 +1,206 @@
+            raise RuntimeError('Learning Rate Plateau threshold mode %s is unknown!')
+        if mode == 'min' and threshold_mode == 'rel':
+            rel_epsilon = 1. - threshold
+            self.monitor_op = lambda a, best: np.less(a, best * rel_epsilon)
```
You can't use numpy. It's not a hard dependency and the same thing can be easily implemented using only standard Python libraries

apaszke replied:
You can't use numpy. It's not a hard dependency and the same thing can be easily implemented using only standard Python libraries

apaszke replied:
You can't use numpy. It's not a hard dependency and the same thing can be easily implemented using only Python libraries

Jiaming-Liu replied:
This part is ported from Keras. But sure, I will remove it.

Jiaming-Liu replied:
Fixed.

Jiaming-Liu replied:
Fixed.

apaszke commented on torch/optim/lr_scheduler.py:
```diff
@@ -0,0 +1,206 @@
+            self.cooldown_counter -= 1
+            self.wait = 0
+
+        if self.monitor_op(current, self.best):
```
maybe rename `moitor_op` to `is_better`? It's really hard to understand what `monitor_op` is supposed to return when called

Jiaming-Liu replied:
Fixed.

apaszke commented on torch/optim/lr_scheduler.py:
```diff
@@ -0,0 +1,206 @@
+        self.best = self.monitor_op.worse
+        self.cooldown_counter = 0
+        self.wait = 0
+        self.lr_epsilon = self.min_lr * 1e-4
```
why is this defined outside of `__init__` and why does it depend on `min_lr`? Epsilon sounds like just a small constant (e.g. 1e-6)

Jiaming-Liu replied:
Agree with the init part. This should be related to min lr. Otherwise the code would be wrong when the user sets a very low lr.

apaszke commented on torch/optim/lr_scheduler.py:
```diff
@@ -0,0 +1,206 @@
+                new_lr = max(new_lr, self.min_lr)
+                param_group['lr'] = new_lr
+                if self.verbose > 0:
+                    print('Epoch %05d: reducing learning rate of group %d to %s.' % (epoch, inx_group, new_lr))
```
I'd replace the loop body with
```python
old_lr = float(param_group['lr'])
new_lr = old_lr * self.factor
if new_nr >= self.min_lr:
    param_group['lr'] = new_lr
    if self.verbose:
        print(...)
```

Jiaming-Liu replied:
This will change the logic.

apaszke reviewed (commented):
Looks good for the most part, but I think some parts could be simplified. Thanks for the PR!

apaszke commented:
Also, tests should go to `test_optim.py`

Jiaming-Liu commented on torch/optim/lr_scheduler.py:
```diff
@@ -0,0 +1,206 @@
+from torch.optim.optimizer import Optimizer
+
+
+class LambdaLR(object):
```
I believe that this part is less complicated than the optimizer. Also, the step() function should be called before each epoch. (or after for ReduceLROnPlateau) This is also different from optimizer.

Jiaming-Liu commented on torch/optim/lr_scheduler.py:
```diff
@@ -0,0 +1,206 @@
+            param_group['lr'] = self.base_lr * self.lr_lambda(epoch)
+
+
+class GroupLambdaLR(object):
```
I am personally not sure about this class. But I think the term base lr is more common than lr base.

Jiaming-Liu replied:
LambdaLR now accepts a scalar or a list of scalar for base_lr :)

Jiaming-Liu commented on torch/optim/lr_scheduler.py:
```diff
@@ -0,0 +1,206 @@
+        >>>     validate(...)
+    """
+
+    def __init__(self, optimizer, base_lr=0.1, gamma=0.1, step_size=30):
```
Yes. I will remove all the default settings.

Jiaming-Liu commented on torch/optim/lr_scheduler.py:
```diff
@@ -0,0 +1,206 @@
+        >>>     validate(...)
+    """
+
+    def __init__(self, optimizer, base_lr=0.1, gamma=0.1, milestones=(10, 20, 30)):
```
IMO, separating them is more natural and easy to understand. Merging these might cause confusion. 
Sort of agree with the verifying part. Is the point about letting the user know their input is wrong without changing it?

Jiaming-Liu commented on torch/optim/lr_scheduler.py:
```diff
@@ -0,0 +1,206 @@
+        patience: number of epochs with no improvement
+            after which learning rate will be reduced.
+        verbose: int. 0: quiet, 1: update messages.
+        mode: one of {min, max}. In `min` mode,
```
I am not quite understanding this comment.  Max returns a value instead of a boolean. Also, making it an option would make it easier to understand.

Jiaming-Liu committed:
Delete default values

Jiaming-Liu committed:
Change Error Msgs

Jiaming-Liu committed:
Change ReduceLROnPlateau

> Monitor_op: renamed to mode_worse. Now created by _make_is_better_op
and contains a lambda exp.
> Removed _MonitorOp class
> Remove numpy dependance.
> verbose: changed to bool

Jiaming-Liu commented on torch/optim/lr_scheduler.py:
```diff
@@ -0,0 +1,206 @@
+                if self.verbose > 0:
+                    print('Epoch %05d: reducing learning rate of group %d to %s.' % (epoch, inx_group, new_lr))
+
+    def in_cooldown(self):
```
I prefer not. We would need to add another function call step_cooldown() if so.

Jiaming-Liu commented on torch/optim/lr_scheduler.py:
```diff
@@ -0,0 +1,206 @@
+                                     threshold_mode=threshold_mode)
+        self.wait = 0
+        self.best = 0
+        assert isinstance(optimizer, Optimizer)
```
Fixed.

Jiaming-Liu commented on torch/optim/lr_scheduler.py:
```diff
@@ -0,0 +1,206 @@
+        self.verbose = verbose
+        self.cooldown = cooldown
+        self.cooldown_counter = 0  # Cooldown counter.
+        self.monitor_op = _MonitorOp(mode=mode, threshold=threshold,
```
Fixed.

Jiaming-Liu committed:
Add unit test

Jiaming-Liu committed:
Revert "Add unit test"

This reverts commit deb0d3da1bd4579bd55b05e6a55824ab4ed56ff5.

Jiaming-Liu committed:
Add unit test

Jiaming-Liu committed:
Pass Flake8

E303 too many blank lines (3)

Jiaming-Liu committed:
ReduceLROnPlateau: Change min_lr to min_lrs

Now ReduceLROnPlateau accepts min_lr as a scalar or a list.

Jiaming-Liu committed:
LambdaLR: Change base_lr to base_lrs

Now LambdaLR and its subclasses accepts base_lr as a scalar or a list.

Jiaming-Liu committed:
Update Doc

Jiaming-Liu committed:
Passing PEP8

Jiaming-Liu committed:
Minor Doc Changes

Jiaming-Liu committed:
Relative import

Post Approved Modify

soumith commented:
@Jiaming-Liu i think this is good to go. You should also add docstrings for LambdaLR and GroupLambdaLR and add references in https://raw.githubusercontent.com/pytorch/pytorch/master/docs/source/optim.rst so that they will show up in documentation as well.
You can test documentation locally by doing this:
```
cd docs
pip install -r requirements.txt
make clean && make html
```
and then locally generated html documentation similar to pytorch.org/docs/ will be in docs/build/html

apaszke committed:
PR fixes

apaszke commented:
@pytorchbot test this please

apaszke commented:
@pytorchbot test this plesae

apaszke commented:
@pytorchbot test this please

colesbury commented on torch/optim/lr_scheduler.py:
```diff
@@ -0,0 +1,234 @@
+
+
+class LambdaLR(object):
+    def __init__(self, optimizer, lr_lambda):
```
If this is public, can you document this class?

colesbury commented on torch/optim/lr_scheduler.py:
```diff
@@ -0,0 +1,234 @@
+            self.lr_lambdas = list(lr_lambda)
+        self.last_epoch = -1
+
+    def step(self, epoch=None):
```
It would also be good to document step

colesbury reviewed (commented):
I think the extensive of lambdas will prevent these classes from being pickled. (They probably could just be instance methods)

Jiaming-Liu committed:
Add param last_epoch

Jiaming-Liu committed:
Update last_epoch in every step(epoch) call

Jiaming-Liu committed:
Fix Pickle-Ablilty Issue

Add a new superclass _LR_Scheduler.

Jiaming-Liu committed:
Minor change

Jiaming-Liu committed:
Rename _LR_Scheduler to _LRScheduler

Jiaming-Liu committed:
Add param base_lrs

But it is not required when last_epoch==-1

Jiaming-Liu committed:
Minor changes: Error Msg

Jiaming-Liu committed:
Minor change: add type check

apaszke commented:
Yes, `get_lr` would be a better name. Also, please remove `base_lr` and use the code I wrote in the commit comments.

Jiaming-Liu committed:
Replace base_lrs

Jiaming-Liu committed:
Add test_lambda_lr

Jiaming-Liu committed:
Minor Change

Jiaming-Liu commented:
This error is weird. Any idea? @apaszke 

> Running optim tests
> ..............F........
> FAIL: test_adagrad_sparse (__main__.TestOptim)
> Traceback (most recent call last):
>   File "test_optim.py", line 285, in test_adagrad_sparse
>     lambda params: optim.Adagrad(params, lr=1e-1)
>   File "test_optim.py", line 103, in _test_rosenbrock_sparse
>     self.assertLessEqual(params.data.dist(solution), initial_dist)
> AssertionError: 0.7290316658655626 not less than or equal to 0.7071067811865476

apaszke commented on torch/optim/lr_scheduler.py:
```diff
@@ -0,0 +1,279 @@
+                    raise KeyError("param 'initial_lr' is not specified "
+                                   "in param_groups[{}] when resuming an optimizer".format(i))
+        self.base_lrs = list(map(lambda group: group['initial_lr'], optimizer.param_groups))
+        self.step(last_epoch + 1)
```
Why are you calling step in the constructor?

Jiaming-Liu replied:
To update the lr in resumed scenario.

apaszke commented on torch/optim/lr_scheduler.py:
```diff
@@ -0,0 +1,279 @@
+            param_group['lr'] = lr
+
+
+class LambdaLR(_LRScheduler):
```
This should be either documented or removed (as Sam said)

Jiaming-Liu replied:
You mean Lambda LR? Sure, I will document it.

apaszke commented on torch/optim/lr_scheduler.py:
```diff
@@ -0,0 +1,279 @@
+class LambdaLR(_LRScheduler):
+    def __init__(self, optimizer, lr_lambda, last_epoch=-1):
+        self.optimizer = optimizer
+        self.base_lrs = list(map(lambda group: group['lr'], optimizer.param_groups))
```
Why is it reading `lr` here? What's wrong with letting `_LRScheduler` figure out what should be the base LRs? All attributes assigned in this constructor (except `lr_lambdas`) will be overwritten by the super constructor, so there's no need to do it.

apaszke reviewed (commented):
Looks good now and should be ready to merge after these final fixes. Can you also add the schedulers to `docs/source/optim.rst`?

apaszke commented:
Also, can you try rebasing on top of master? The test might have been fixed in some other commit.

Jiaming-Liu commented:
Now it seems like a good time to update the documentation. I will have it done within a week. Thanks for the reviews!

Jiaming-Liu commented:
Rebasing doesn't help the error :(. Any ideas?

Jiaming-Liu committed:
Remove redundant lr assignment

Jiaming-Liu committed:
Minor Change in test

Jiaming-Liu committed:
Minor Change in Test

thomasjpfan commented:
The test looks like it's been fixed at 368ecb47f97c6dc2d57c487bda451df71148e488. Rebasing on top of master fixes the error.

Jiaming-Liu commented:
umm... something is wrong with the rebasing

Jiaming-Liu committed:
Update docs

Jiaming-Liu committed:
Update optim.rst

Jiaming-Liu committed:
Minor doc fix

soumith commented:
@pytorchbot test this please

szagoruyko commented:
As far as I see there is only one optimizer being kept, so on learning rate drop all other parameters are also kept. How would one add momentum resetting on learning rate drops in SGD?

Jiaming-Liu commented:
@soumith Kindly mention this pr in some release note to increase visibility.

soumith commented:
hey jiaming. I'm really sorry for missing this commit in the release notes. It looks like I missed 4 commits by mistake. I've updated the release notes now, and I've made a note for myself to check if repeating the note about learning rate schedules will be appropriate for the next release as well (to increase visibility)

jtoy commented:
so this is in the new release 0.2.0 ? great!

FuriouslyCurious commented:
@soumith  My post on PyTorch forum about LR schedules is still getting more likes every week, so I think people are not aware of this PR.  You should consider megaphoning this PR in future release notes. 

https://discuss.pytorch.org/t/adaptive-learning-rate/320/10

victorhcm commented on torch/optim/lr_scheduler.py:
```diff
@@ -0,0 +1,305 @@
+        >>> lambda2 = lambda epoch: 0.95 ** epoch
+        >>> scheduler = LambdaLR(optimizer, lr_lambda=[lambda1, lambda2])
+        >>> for epoch in range(100):
+        >>>     scheduler.step()
```
Sorry for resurrecting this, @Jiaming-Liu, I just wanted to check with you first. In the docstrings, should we change `scheduler.step()` to `scheduler.step(epoch)`, such as in the tests? The current example will update the learning rate for every epoch, which might be misleading in the documentation.
